File: go/meetup/meetup.go
package meetup

import (
	"time"
)

const testVersion = 3

type WeekSchedule int

const (
	First WeekSchedule = iota
	Second
	Third
	Fourth
	Last
	Teenth
)

func Day(ws WeekSchedule, weekday time.Weekday, month time.Month, year int) int {
	date := time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
	switch ws {
	case First:
		for date.Weekday() != weekday {
			date = date.AddDate(0, 0, 1)
		}
		return date.Day()
	case Second:
		return Day(First, weekday, month, year) + 7
	case Third:
		return Day(First, weekday, month, year) + 14
	case Fourth:
		return Day(First, weekday, month, year) + 21
	case Last:
		date = date.AddDate(0, 1, -1) // Go the the last day of the month
		for date.Weekday() != weekday {
			date = date.AddDate(0, 0, -1)
		}
		return date.Day()

	case Teenth:
		for date.Weekday() != weekday || !(12 < date.Day() && date.Day() < 20) {
			date = date.AddDate(0, 0, 1)
		}
		return date.Day()
	}
	return 0
}

File: haskell/meetup/src/Meetup.hs
module Meetup
  ( Weekday(..)
  , Schedule(..)
  , meetupDay
  ) where

import qualified Data.Time.Calendar          as D

import           Data.List                   (nub)
import           Data.Time.Calendar.WeekDate (toWeekDate)

data Weekday
  = Monday
  | Tuesday
  | Wednesday
  | Thursday
  | Friday
  | Saturday
  | Sunday
  deriving (Enum)

data Schedule
  = First
  | Second
  | Third
  | Fourth
  | Last
  | Teenth
  deriving (Eq)

meetupDay :: Schedule -> Weekday -> Integer -> Int -> D.Day
meetupDay schedule weekday year month =
  head $ filter (matchWeekDay . toWeekDate) (candidates schedule year month)
  where
    matchWeekDay (_, _, w) = (fromEnum weekday + 1) == w

candidates :: Schedule -> Integer -> Int -> [D.Day]
candidates schedule year month
  | schedule == First = dayList [1 .. 7]
  | schedule == Second = dayList [8 .. 14]
  | schedule == Third = dayList [15 .. 21]
  | schedule == Fourth = dayList [22 .. 28]
  | schedule == Last = take 7 $ nub $ dayList [31,30 .. 22]
  | schedule == Teenth = dayList [13 .. 19]
  where
    dayList = map (D.fromGregorian year month)
